src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -151,17 +151,73 @@ void ClusterTaskManager::TasksUnblocked(const std::vector<TaskID> ready_ids) {
   }
 }
 
+void ClusterTaskManager::TaskFinished(const TaskID &task_id) {
+  auto iter = running_tasks_.find(task_id);
+  if (iter != running_tasks_.end()) {
+    running_tasks_.erase(iter);
+  } else {
+    // This shouldn't happen (unless the worker resends a ReturnWorker request?).
+    RAY_LOG(DEBUG) << "Received TaskDone for unknown task.";
```

Comment:
Consider making this a check.

Reply:
I probably should've marked this as a TODO.

Do we have a guarantee that the worker won't retry? From a quick glance, I didn't see a retry mechanism in the direct task transport

Reply:
The worker should not retry.